File: cousinsoftree.java
public /**
* Definition for a binary tree node.
* public class TreeNode {
*     int val;
*     TreeNode left;
*     TreeNode right;
*     TreeNode() {}
*     TreeNode(int val) { this.val = val; }
*     TreeNode(int val, TreeNode left, TreeNode right) {
*         this.val = val;
*         this.left = left;
*         this.right = right;
*     }
* }
*/
// time complexity:- o(n)
//space complexity :- height of the tree , o(h)
class Solution {
   
       TreeNode x_parent; TreeNode y_parent;
       int x_depth; int y_depth;
   
   public boolean isCousins(TreeNode root, int x, int y) 
   {

       if(root == null) return false;
       dfs(root, x, y, 0,null);
       return (x_parent != y_parent) && (x_depth == y_depth);   
       
   }
   private void dfs(TreeNode root, int x, int y, int depth, TreeNode parent)
   {
       //base

       if(root == null) return;

       
       if(root.val == x)
       {
          
           x_parent = parent;
           x_depth = depth;
           
           
       }
       if(root.val == y)
       {
           y_parent = parent;
           y_depth = depth;
           
       }

       
       dfs(root.left, x, y, depth +1, root);
       dfs(root.right, x, y, depth +1, root);
   }
}

File: CousinsOfTree.py
#Time complexity: O(n)
#Space complexity: O(h)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    xParent=None
    yParent=None
    xLevel=0
    yLevel=0
    def isCousins(self, root, x: int, y: int) -> bool:
        self.xParent=None
        self.yParent=None
        self.xLevel=0
        self.yLevel=0
        self.findParentAndLevel(root,x,y,0,None)
        return (self.xLevel==self.yLevel and self.xParent!=self.yParent)
    def findParentAndLevel(self,root,x,y,level,parent):
        if not root:
            return
        if root.val==x:
            self.xParent=parent
            self.xLevel=level
            return
        if root.val==y:
            self.yParent=parent
            self.yLevel=level
            return
        self.findParentAndLevel(root.left,x,y,level+1,root)
        self.findParentAndLevel(root.right,x,y,level+1,root)
        
            
            
        